aten/src/ATen/native/cuda/Shape.cu

```diff
@@ -0,0 +1,399 @@
+          d_inputs,
+          stackInputs,
+          batchCounter * sizeof(CatArrInputTensor<scalar_t, unsigned int>),
+          cudaMemcpyHostToDevice,
+          stream.stream()));
+      THCudaHostRecord(state, stackInputs);
+    }
+
+    // Next, let's consider how we set our kernel launch parameters.
+    // We borrow from THCApply, which the kernel's internal indexing
+    // is based on.
+    dim3 applyBlock = dim3(32*16);
+
+    //Get grid where x dim fills half gpu and y dim is number of tensors.
+    //This will have cating two tensors fill the entire grid, but prevent
+    //many threads from needlessly load meta data if their sizes is small.
+    dim3 catGrid;
+    getCatGrid(state, batchCounter, catGrid);
+
+
+    // Template Declarations for dim = 1, 2, 3, 4
+#define HANDLE_CASE(DIMS) \
+    CatArrayBatchedCopy<scalar_t, unsigned int, DIMS><<<\
+        catGrid, applyBlock, 0, stream.stream()>>>(\
+            data, d_inputs, param, dimension, param.outputStride[dimension]);
+    switch (nDims) {
+      case 1:
+        HANDLE_CASE(1);
+        break;
+      case 2:
+        HANDLE_CASE(2);
+        break;
+      case 3:
+        HANDLE_CASE(3);
+        break;
+      case 4:
+        HANDLE_CASE(4);
+        break;
+    }
+#undef HANDLE_CASE
+    THCudaCheck(cudaGetLastError());
+  }
+  THCudaFree(state, d_inputs);
+}
+
+} // namespace
+
+Tensor cat_cuda(TensorList inputs, int64_t dimension) {
+  Tensor out = at::empty({0}, inputs.front().options());
+  cat_out_cuda(out, inputs, dimension);
+  return out;
+}
+
+Tensor& cat_out_cuda(Tensor& out, TensorList inputs, int64_t dimension) {
+  THCState *state = at::globalContext().getTHCState();
+
+  // previously, size [0] tensors were the only possible empty tensors; thus, it
+  // wasn't possible to cat empty tensors unless all the other tensors were
+  // 1-dimensional, so we allowed these tensors to be "skipped".  We maintain
+  // this behavior for backwards compatibility, but only for this specific size
+  // (i.e. other empty sizes are not skipped).
+  // FIXME: warn if this is the case
+  auto should_skip = [](const Tensor &t) {
+    return t.dim() == 1 && at::native::size(t, 0) == 0;
+  };
+  bool hasSkippedInput = false;
+
+  const Tensor *notSkippedTensor = NULL;  // non-owning reference
+  int nDims = 0;
+
+  // Inputs cannot alias the output tensor
+  for (int i = 0; i < inputs.size(); i++) {
+    auto lap = at::get_overlap_status(out, inputs[i]);
+    TORCH_CHECK(lap != at::MemOverlapStatus::PARTIAL &&
+                lap != at::MemOverlapStatus::FULL,
+                "unsupported operation: the input tensors cannot refer to any "
+                "of the output memory locations. Found overlap in input "
+                "tensor ", i);
+  }
+
+  for (int i = 0; i < inputs.size(); i++)
+  {
+    if (should_skip(inputs[i])) {
+      hasSkippedInput = true;
+      continue;
+    }
+    nDims = inputs[i].dim();
+    notSkippedTensor = &inputs[i];
```

Comment:
you can break after finding the first notSkippedTensor

Reply:
only if `hasSkippedInput` has been already set. (Because we later check whether `hasSkippedInput` has been set for any tensor.) It feels to me that the loss in readability is worse than gain in performance, but please let me know if you think otherwise.